File: service/battle/lockstep/core/pending_commnads.lua
local skynet = require "skynet"
local class	   =  require "class"
local PendingCommnads   = class("PendingCommnads")

local constant = require "lockstep.constant"

function PendingCommnads:initialize()
    self.NumberOfPlayers = constant.NumberOfPlayers

    self.CurrentCommands = self:new_comands(self.NumberOfPlayers)
    self._nextCommands = self:new_comands(self.NumberOfPlayers)
    self._nextNextCommands = 0
    self._nextNextNextCommands = 0

    self._currentCommandsCount = 0
    self._nextCommandsCount = 0
    self._nextNextCommandsCount = 0
    self._nextNextNextCommandsCount = 0
end

function PendingCommnads:new_comands(len)
    -- local command = require "lockstep.command.command"
    local r = {}
    for i = 1, len do 
        table.insert(r, 0)
    end
    return r
end

function PendingCommnads:NextTurn()
    -- Finished processing this turns actions - clear it
    --last turn's actions is now this turn's actions
    self.CurrentCommands = self._nextCommands
    self._currentCommandsCount = self._nextCommandsCount

    --last turn's next next actions is now this turn's next actions
    self._nextCommands = self._nextNextCommands
    self._nextCommandsCount = self._nextNextCommandsCount

    self._nextNextCommands = self._nextNextNextCommands
    self._nextNextCommandsCount = self._nextNextNextCommandsCount

    --set NextNextNextActions to the empty list
    self._nextNextNextCommands = self:new_comands(self.NumberOfPlayers)
    self._nextNextNextCommandsCount = 0    
end

function PendingCommnads:AddCommand(cmd, playerId, currentLockStepTurn, cmdsLockStepTurn)
    --add cmd for processing later
    if cmdsLockStepTurn == currentLockStepTurn + 1 then
        --if action is for next turn, add for processing 3 turns away
        local c = self._nextNextNextCommands[playerId]
        if type(c) == "table" then
            -- TODO: Error Handling
            ERROR("Recieved multiple actions for player " + playerId + " for turn "  + cmdsLockStepTurn)
        end

        self. _nextNextNextCommands[playerId] = cmd
        self._nextNextNextCommandsCount = self._nextNextNextCommandsCount + 1
    elseif cmdsLockStepTurn == currentLockStepTurn then
        --if recieved action during our current turn
        --add for processing 2 turns away
        local c = self._nextNextCommands[playerId]
        if type(c) == "table" then 
            -- TODO: Error Handling
            ERROR("Recieved multiple actions for player " + playerId + " for turn "  + cmdsLockStepTurn)
        end

        self._nextNextCommands[playerId] = cmd
        self._nextNextCommandsCount = self._nextNextCommandsCount + 1
    elseif cmdsLockStepTurn == currentLockStepTurn - 1 then
        --if recieved action for last turn
        --add for processing 1 turn away
        local c = self._nextCommands[playerId]
        if type(c) == "table" then
            --TODO: Error Handling
            ERROR("Recieved multiple actions for player " + playerId + " for turn "  + cmdsLockStepTurn)
        end
        self._nextCommands[playerId] = cmd
        self._nextCommandsCount = self._nextCommandsCount + 1
    else 
        --TODO: Error Handling
        ERROR(" Unexpected lockstepID recieved : " + cmdsLockStepTurn)
    end
end

function PendingCommnads:ReadyForNextTurn() 
    if self._nextNextCommandsCount == self.NumberOfPlayers then
        -- if this is the 2nd turn, check if all the actions sent out on the 1st turn have been recieved
        if constant.LockStepTurnId == constant.FirstLockStepTurnID + 1 then
            return true;
        end
    
        -- Check if all Actions that will be processed next turn have been recieved
        if self._nextCommandsCount == self.NumberOfPlayers then
            return true
        end
    end

    -- if this is the 1st turn, no actions had the chance to be recieved yet
    if constant.LockStepTurnId == constant.FirstLockStepTurnID then
        return true
    end
    --if none of the conditions have been met, return false
end

function PendingCommnads:WhosNotReady() 
    if self._nextNextCommandsCount == self.NumberOfPlayers then
        -- if this is the 2nd turn, check if all the actions sent out on the 1st turn have been recieved
        if constant.LockStepTurnId == constant.FirstLockStepTurnID + 1 then
            return
        end
    
        -- Check if all Actions that will be processed next turn have been recieved
        if self._nextCommandsCount == self.NumberOfPlayers then
            return
        else
            return self:DoWhosNotReady (self._nextCommands, self._nextCommandsCount)
        end
    
    elseif(constant.LockStepTurnId == constant.FirstLockStepTurnID) then
        --if this is the 1st turn, no actions had the chance to be recieved yet
        return
    else
        return self:DoWhosNotReady (self._nextNextCommands, self._nextNextCommandsCount)
    end
end

function PendingCommnads:DoWhosNotReady(actions, count) 
    if(count < constant.NumberOfPlayers) then
        local notReadyPlayers = array_new(self.NumberOfPlayers - count, 0)
        local index = 1
        for playerId = 1,  self.NumberOfPlayers do
            local c = self.actions[playerId]
            if not type(c) == "table" then
                notReadyPlayers[index] = playerId
                index = index + 1
            end
        end
        return notReadyPlayers
    end
end


return PendingCommnads

File: tools/skynet-client/lockstep/Core/PendingCommnads.cs
namespace Skynet.DotNetClient.LockStep
{
    using Utils.Logger;
    
    public class PendingCommnads
    {
        public Command[] CurrentCommands;

        //incase other players advance to the next step and send their action before we advance a step
        private Command[] _nextCommands;
        private Command[] _nextNextCommands;
        private Command[] _nextNextNextCommands;

        private int _currentCommandsCount;
        private int _nextCommandsCount;
        private int _nextNextCommandsCount;
        private int _nextNextNextCommandsCount;
        
        private readonly LockStepManager _lockStepManager;


        public PendingCommnads()
        {
            _lockStepManager = LockStepManager.Instance;
            
            
            CurrentCommands = new Command[_lockStepManager.NumberOfPlayers];
            _nextCommands = new Command[_lockStepManager.NumberOfPlayers];
            _nextNextCommands = new Command[_lockStepManager.NumberOfPlayers];
            _nextNextNextCommands = new Command[_lockStepManager.NumberOfPlayers];
		
            _currentCommandsCount = 0;
            _nextCommandsCount = 0;
            _nextNextCommandsCount = 0;
            _nextNextNextCommandsCount = 0;
        }
        
        public void NextTurn() 
        {
            //Finished processing this turns actions - clear it
            for(int i=0; i<CurrentCommands.Length; i++) 
            {
                CurrentCommands[i] = null;
            }
            Command[] swap = CurrentCommands;
		
            //last turn's actions is now this turn's actions
            CurrentCommands = _nextCommands;
            _currentCommandsCount = _nextCommandsCount;
		
            //last turn's next next actions is now this turn's next actions
            _nextCommands = _nextNextCommands;
            _nextCommandsCount = _nextNextCommandsCount;
		
            _nextNextCommands = _nextNextNextCommands;
            _nextNextCommandsCount = _nextNextNextCommandsCount;
		
            //set NextNextNextActions to the empty list
            _nextNextNextCommands = swap;
            _nextNextNextCommandsCount = 0;
        }

        public void AddCommand(Command cmd, int playerId, int currentLockStepTurn, int cmdsLockStepTurn)
        {
            //add cmd for processing later
            if(cmdsLockStepTurn == currentLockStepTurn + 1) 
            {
                //if action is for next turn, add for processing 3 turns away
                if(_nextNextNextCommands[playerId] != null) 
                {
                    //TODO: Error Handling
                    SkynetLogger.Error(Channel.LockStep,"Recieved multiple actions for player " + playerId + " for turn "  + cmdsLockStepTurn);
                }
                _nextNextNextCommands[playerId] = cmd;
                _nextNextNextCommandsCount++;
            } 
            else if(cmdsLockStepTurn == currentLockStepTurn) 
            {
                //if recieved action during our current turn
                //add for processing 2 turns away
                if(_nextNextCommands[playerId] != null) 
                {
                    //TODO: Error Handling
                    SkynetLogger.Error(Channel.LockStep,"Recieved multiple actions for player " + playerId + " for turn "  + cmdsLockStepTurn);
                }
                _nextNextCommands[playerId] = cmd;
                _nextNextCommandsCount++;
            } 
            else if(cmdsLockStepTurn == currentLockStepTurn - 1) 
            {
                //if recieved action for last turn
                //add for processing 1 turn away
                if(_nextCommands[playerId] != null) {
                    //TODO: Error Handling
                    SkynetLogger.Error(Channel.LockStep,"Recieved multiple actions for player " + playerId + " for turn "  + cmdsLockStepTurn);
                }
                _nextCommands[playerId] = cmd;
                _nextCommandsCount++;
            } 
            else 
            {
                //TODO: Error Handling
                SkynetLogger.Error(Channel.LockStep," Unexpected lockstepID recieved : " + cmdsLockStepTurn);
            }
        }
        
        public bool ReadyForNextTurn() 
        {
            if(_nextNextCommandsCount == _lockStepManager.NumberOfPlayers) {
                //if this is the 2nd turn, check if all the actions sent out on the 1st turn have been recieved
                if(_lockStepManager.LockStepTurnId == LockStepManager.FirstLockStepTurnID + 1) {
                    return true;
                }
			
                //Check if all Actions that will be processed next turn have been recieved
                if(_nextCommandsCount == _lockStepManager.NumberOfPlayers) {
                    return true;
                }
            }
		
            //if this is the 1st turn, no actions had the chance to be recieved yet
            if(_lockStepManager.LockStepTurnId == LockStepManager.FirstLockStepTurnID) {
                return true;
            }
            //if none of the conditions have been met, return false
            return false;
        }
        
        public int[] WhosNotReady() 
        {
            if(_nextNextCommandsCount == _lockStepManager.NumberOfPlayers) {
                //if this is the 2nd turn, check if all the actions sent out on the 1st turn have been recieved
                if(_lockStepManager.LockStepTurnId == LockStepManager.FirstLockStepTurnID + 1) {
                    return null;
                }
			
                //Check if all Actions that will be processed next turn have been recieved
                if(_nextCommandsCount == _lockStepManager.NumberOfPlayers) {
                    return null;
                }else {
                    return WhosNotReady (_nextCommands, _nextCommandsCount);
                }
			
            } else if(_lockStepManager.LockStepTurnId == LockStepManager.FirstLockStepTurnID) {
                //if this is the 1st turn, no actions had the chance to be recieved yet
                return null;
            } else {
                return WhosNotReady (_nextNextCommands, _nextNextCommandsCount);
            }
        }
	
        private int[] WhosNotReady(Command[] actions, int count) 
        {
            if(count < _lockStepManager.NumberOfPlayers) {
                var notReadyPlayers = new int[_lockStepManager.NumberOfPlayers - count];
			
                var index = 0;
                for(var playerId = 0; playerId < _lockStepManager.NumberOfPlayers; playerId++) {
                    if(actions[playerId] == null) {
                        notReadyPlayers[index] = playerId;
                        index++;
                    }
                }
			
                return notReadyPlayers;
            } else {
                return null;
            }
        }
        
    }
}